src/main/scala/lms/core/backend.scala

```diff
@@ -322,51 +322,88 @@ class GraphBuilder {
+  // FIXME(feiw) Dig further to see if/why lambda_forward or None cases are correct
+  // FIXME(feiw) in the conditional case, the handling of result is still wrong.
+  def getLambdaLatentEffect(f: Exp): ((Set[Exp], Set[Exp]),(Set[Int], Set[Int]), Option[Exp]) = findDefinition(f) match {
+      case Some(Node(_, "λ", List(b:Block), _)) =>
+        getEffKeysWithParam(b)
+      case Some(Node(_, "λforward", _, _)) => // what about doubly recursive?
+        ((Set[Exp](), Set[Exp](Const("CTRL"))), (Set[Int](), Set[Int]()), None)
+      case None => // FIXME: function argument? fac-01 test used for recursive function...
+        ((Set[Exp](), Set[Exp](Const("CTRL"))), (Set[Int](), Set[Int]()), None)
+      case Some(Node(_, "@", (f: Sym)+:args, _)) =>
+        val ((rk, wk), Some(f_res)) = getApplyLatentEffect(f, args: _*)
+        val ((rk2, wk2), (prk2, pwk2), f_res_res) = getLambdaLatentEffect(f_res)
+        ((rk ++ rk2, wk ++ wk2), (prk2, pwk2), f_res_res)
+      case Some(Node(_, "?", c::Block(ins, out, ein, eout)::Block(ins2, out2, ein2, eout2)::Nil, _)) =>
+        val ((rk, wk), (prk, pwk), _) = getLambdaLatentEffect(out)
+        val ((rk2, wk2), (prk2, pwk2), _) = getLambdaLatentEffect(out2)
+        ((rk ++ rk2, wk ++ wk2), (prk ++ prk2, pwk ++ pwk2), None) // FIXME(feiw)
+      case Some(e) =>
+        ??? // FIXME what about @, ?, array_apply => conservative write on all args?
+  }
 
-      // We also need to add Const effect:
-      (reads ++ freads.filter(_.isInstanceOf[Const]), writes ++ fwrites.filter(_.isInstanceOf[Const]))
-    case _ =>
-      val reads = new mutable.HashSet[Exp]
-      val writes = new mutable.HashSet[Exp]
-      for (x <- xs) {
-        val (ref, wef) = getLatentEffect(x)
-        reads ++= ref
-        writes ++= wef
-      }
-      (reads, writes)
+  // The sub function for getting latent effects for function application
+  // 1. get the lambda latent effects from the function
+  // 2. update the lambda latent effects (replace parameters with arguments)
+  // The sub functon also returns the result of the function because the result might
+  //    be another function that is applied (we need to get its latent effects).
+  def getApplyLatentEffect(f: Sym, args: Def*): ((Set[Exp], Set[Exp]), Option[Exp]) = {
+    val (reads, writes) = getLatentEffect(args: _*)
+    val ((freads, fwrites), (preads, pwrites), res) = getLambdaLatentEffect(f)
+    // For @ we need to replace the effect on parameters to the actual arguments.
+    // the asInstanceOf seems unsafe at first glance. However, it is not a problem since a standalone block
+    // should never be an argument in function application.
+    ((reads ++ freads ++ preads.map(args(_).asInstanceOf[Exp]), writes ++ fwrites ++ pwrites.map(args(_).asInstanceOf[Exp])), res)
+  }
+
+  // The main function for getting latent effects
+  // Latent effects are effects in Blocks (conditional, loop, lambda)
```

Comment:
Actually I am not so sure about the definition of Latent Effect. Initially I thought it should be `effects of functions`, which are latent because they show up in applications rather than function definition. 
However, the code seems to treat blocks of conditionals and loops as `latent effects` too, which is not that accurate IMO since they are not in any way `latent`. 

What do you think a proper description of latent effect is?